### ISSUE TITLE
PEP 604 Support

### ISSUE THREAD

**Issue by AngellusMortis** [fastapi/typer#348]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pathlib import Path
import typer


def main(output: Path | None):
    typer.guess_name(f"Hello {name}")


if __name__ == "__main__":
    typer.run(main)
```

### Description

Using an argument with the type of `pathlib.Path | None` (as per [PEP 604](https://www.python.org/dev/peps/pep-0604/)) raise the error:

```
Type not yet supported: pathlib.Path | None
```

### Wanted Solution

`| None` types should work instead of having to need to use `Optional[]`.

### Wanted Code

```python
from pathlib import Path
import typer


def main(output: Path | None):
    typer.echo(output)


if __name__ == "__main__":
    typer.run(main)
```

### Operating System

Linux

### Typer Version

0.4.0

### Python Version

3.10.1

**Comment by navhits** [comment]
I'm thinking if you can do this `output: Path = None`? When you have `Path | None`, the function expects either `Path` or `None` to be passed. Plus I believe Typer is tested for up to Python 3.9. This means most of the features from Python 3.10 might not work as expected. Let's just say certain recursive dependencies used by Typer also might have similar troubles.

**Comment by SRv6d** [comment]
+1
Even using Python 3.10 without the `__future__`  imports PEP 604 isn't supported.
@tiangolo Are there any plans to support this in the near future?

**Comment by navhits** [comment]
Could this be more of a Pydantic issue?

**Comment by SRv6d** [comment]
@navhits I'm not exactly sure what you mean. Pydantic supports PEP 604 type hints.

**Comment by navhits** [comment]
If that's the case, then try building from [python3.10 branch](https://github.com/tiangolo/typer/tree/python3.10). In my understanding 604 is a 3.10 focussed enhancement.

**Comment by navhits** [comment]
Ok. There's a [0.4.1](https://github.com/tiangolo/typer/releases/tag/0.4.1) release already. Try this version @SRv6d

**Comment by SRv6d** [comment]
I've been on that version since march.

**Comment by Sighery** [comment]
Seems this is still broken as of 0.7.0, even though that release is supposed to support Python 3.11 (and therefore I assumed these new type annotations included with Python3.10)

**Comment by aberres** [comment]
Yeps, running into this as well.

As we have `pyupgrade` in place enforcing PEP 604, I am in an awkward situation 🤔

**Comment by AngellusMortis** [comment]
> As we have pyupgrade in place 

Make them standalone types and pyupgrade will not change it.

```
PathOrNone = Path | None

def main(output: PathOrNone):
    typer.guess_name(f"Hello {name}")
```

**Comment by Sighery** [comment]
> > As we have pyupgrade in place
> 
> Make them standalone types and pyupgrade will not change it. Or exclude the file from pyupgrade (your CLI interface is separated from the rest of your app, right?).
> 
> ```python
> PathOrNone = Path | None
> 
> def main(output: PathOrNone):
>     typer.guess_name(f"Hello {name}")
> ```

Last I tried, making them standalone types would still give errors in `typer`, as it just doesn't support the new union notation at all. If you pass it a standalone type, it'll just go through an extra hop to get the types that are in an union in the standalone type, and then fail again because it doesn't understand `|`

**Comment by AngellusMortis** [comment]
Sorry, bad example, I just woke up. 😂 

```python
PathOrNone = Optional[Path]

def main(output: PathOrNone):
    typer.guess_name(f"Hello {name}")
```

pyupgrade should not change `Optional[]` type aliases.

**Comment by aberres** [comment]
> ```python
> PathOrNone = Optional[Path]
> ```

Nice one, thanks!

**Comment by johnthagen** [comment]
Since the PR doesn't link to this issue, I'll link it here. This PR appears to be trying to fix this issue:

- https://github.com/tiangolo/typer/pull/522

**Comment by johnthagen** [comment]
For visibility, the current open PR for this issue is

- #676

**Comment by svlandeg** [comment]
Confirmed that either #676 or #548 will fix this issue!

**Comment by Spenhouet** [comment]
Pydantic also has this issue and supports the union type as of pydantic 2.5. Not sure if this requires typer to support pydantic >=2.5?

**Comment by tiangolo** [comment]
Thanks everyone for the feedback! :coffee: 

This should have been fixed by https://github.com/fastapi/typer/pull/548

It is now available in Typer `0.12.4` :rocket:
